crates/typst-utils/src/format.rs

```diff
@@ -0,0 +1,187 @@
+//! Different formatting functions.
+
+use std::time::Duration;
+
+/// Returns value with `n` digits after floating point where `n` is `precision`.
+/// Standard rounding rule applies (if `n+1`th digit >= 5 then round up).
+/// If `value` is +/- infinity or NaN returns `value`.
+pub fn round_with_precision(value: f64, precision: u8) -> f64 {
+    if value.is_infinite()
+        || value.is_nan()
+        // Integers that big can't have fractional part (mantissa is already full).
+        || value.abs() >= (1_i64 << 53) as f64
+        // Binary64 format can only precisely represent up to log_10(2^53) digits.
+        || precision >= 17
+    {
+        return value;
+    }
+    let offset = 10_f64.powi(precision.into());
+    if !(value * offset).is_finite() {
+        return value;
+    }
+    (value * offset).round() / offset
+}
+
+/// Returns number of `(days, hours, minutes, seconds, milliseconds, microseconds)`.
+fn get_duration_parts(duration: &Duration) -> (u16, u8, u8, u8, u16, u16) {
+    // In practice we probably don't need nanoseconds.
+    let micros = duration.as_micros();
+    let (millis, micros) = (micros / 1000, (micros % 1000) as u16);
+    let (sec, millis) = (millis / 1000, (millis % 1000) as u16);
+    let (mins, sec) = (sec / 60, (sec % 60) as u8);
+    let (hours, mins) = (mins / 60, (mins % 60) as u8);
+    let (days, hours) = ((hours / 24) as u16, (hours % 24) as u8);
+    (days, hours, mins, sec, millis, micros)
+}
+
+/// Format string using `days`, `hours`, `minutes`, `seconds`.
+fn format_dhms(days: u16, hours: u8, minutes: u8, seconds: u8) -> String {
+    match (days, hours, minutes, seconds) {
+        (0, 0, 0, s) => format!("{s:2} s"),
+        (0, 0, m, s) => format!("{m:2} m {s:2} s"),
+        (0, h, m, s) => format!("{h:2} h {m:2} m {s:2} s"),
+        (d, h, m, s) => format!("{d:3} d {h:2} h {m:2} m {s:2} s"),
+    }
+}
+
+/// Format string starting with number of days and going bigger from there.
+pub fn time_starting_with_seconds(duration: &Duration) -> String {
+    let (days, hours, minutes, seconds, _, _) = get_duration_parts(duration);
+    format_dhms(days, hours, minutes, seconds)
+}
+
+/// Format string starting with number of milliseconds and going bigger from there.
+/// `precision` is how many digits of microseconds from floating point to the right
+/// will be preserved. Note that this function will always remove all trailing zeros
+/// for microseconds.
+pub fn time_starting_with_ms_with_precision(
```

Comment:
This function's name is quite long, and a bit misleading (precision of what? - only the documentation can clarify this). I think we can drop the `with_precision` part.